Implement new playlist/error API

ryanheise opened the pull request:
Closes #1421 
Fixes #353
Fixes #390

To test, please comment/uncomment the lines in each `pubspec.yaml` as indicated.

ryanheise committed:
Merge branch 'fix/swiftpm_macos' into minor

ryanheise committed:
New playlist and error APIs.

ryanheise committed:
Add iOS/bg implementations for new playlist/error API

ryanheise committed:
Fix bugs in error handling.

ryanheise committed:
Add errorStream.

ryanheise committed:
Convert state diagram to use new playlist API.

Chaphasilor commented:
Thanks for working on this!

Will there be a (rough) migration guide for this change?  
And since it's a breaking change, will `ConcatenatingAudioSource` be removed immediately or just deprecated?  
Lastly, is the usage of `ShuffleOrder` affected by this in any significant way?

ryanheise committed:
Bump min flutter version to 3.27.0, AGP to 8.5.2.

ryanheise committed:
Sync sdk with flutter version.

ryanheise committed:
Add .cxx to .gitignore

ryanheise commented:
`ConcatenatingAudioSource` won't be removed immediately, just deprecated. `ShuffleOrder` is not affected.

Regarding migration, I wasn't planning on creating a migration guide, since but let me know your experience migrating. If you find it non-obvious, then I'll add a migration guide.

ryanheise committed:
Bump version to 4.5.0 for just_audio_platform_interface.

ryanheise committed:
Update example to use just_audio_media_kit

ryanheise committed:
Bump audio_session dep to 0.2.0.

ryanheise committed:
Update background to audio_session 0.2.0

hpoul committed:
Allow audio_session 0.2.0

audio_session 0.2.x does not introduce any breaking changes to the API, but only to the iOS setup. Hence no code changes are required.

ryanheise committed:
Merge branch 'patch-1' of github.com:hpoul/just_audio into hpoul-patch-1

ryanheise committed:
Merge branch 'hpoul-patch-1' into feature/playlist

ryanheise committed:
Update example to use latest just_audio_media_kit.

ryanheise committed:
Relax audio_session constraint on just_audio_background.

Jshewmaker commented on just_audio/lib/just_audio.dart:
```diff
@@ -1847,11 +2041,17 @@ class IcyMetadata {
 
   /// The index of the current source in the sequence.
-  final int currentIndex;
+  int? get currentIndex =>
```
@ryanheise doesn't this change break existing functionality? Now when a audio player is loading for the first time, it always emits 0 first, even if the correct track is a different index. We would rather the stream emit null until the correct value is ready to be emitted.

Jshewmaker replied:
player.currentIndexStream now always emits 0 even when it shouldnt on first load
